content/articles/oop-in-java-abstraction-and-encapsulation/index.md

```diff
@@ -0,0 +1,456 @@
+ 
+}
+ 
+}
+ 
+```
+ 
+Now you see, we’ve created a new dog with the name `Bingo`. You can go on to create as many dogs with different states or attributes.
+ 
+Here is the output:
+ 
+  ```
+  Dog running
+ 
+  Is the dog running:? true
+ 
+  Sleeping
+ 
+  is the dog sleeping:? true
+ 
+  Dog barking...
+ 
+  Is the dog barking:? true
+ 
+  Eating
+ 
+  Is the dog eating:? true
+ 
+  The dog name is: Bingo
+  ```
+ 
+Now that we’ve got this out of the way let's talk about abstraction and encapsulation.
+ 
+### Abstraction 
+Who could have imagined that abstraction will be a thing in programming? Who? But right now, you can’t talk about OOP without mentioning abstraction. It is a significant building block of OOP. So then, what is abstraction? This section will explain what abstraction is and how to use it as a Java developer.
+
+In abstraction, only the useful attributes of an object are on display or accessible. The low-level details are *hidden*. For example, imagine you’re typing on your computer. You’re interested in punching the keys believing that it produces results on your screen. You know nothing about the mechanism that produces the result, and I’m sure you haven’t even thought about it yet. Have you?
+ 
+Here's another example to make things clearer: imagine car drivers. They are only interested in starting, moving, and stopping the car, speed, or changing gear and brake. They are not interested in the mechanism that allows them to do this. In abstraction, we hide the implementation details from the user while exposing the functionality.
+ 
+This way, the user will only know what a function (method in java) does but not how it does it.
+ 
+### How to implement Abstraction in Java
+In Java, abstraction is implemented using either of these two ways:
+- By using the non-access modifier *abstract* keyword or
+- Interface
+ 
+#### Working with the Abstract Keyword
+You can only use the abstract access modifier keyword with classes and methods but not with variables. Methods with the keyword *abstract* are methods without implementation.
+ 
+To implement an abstract class in Java, use the abstract keyword, e.g., ```public abstract class Circle```, where Circle is the class name. Let's illustrate this using this code snippet:
+ 
+```java
+ 
+public abstract class Dog {
+ 
+ public abstract void run();
+ 
+}
+ 
+```
+ 
+With the above code snippet, we've created an abstract class and an abstract method. One pertinent feature of abstract class is that objects cannot be instantiated or created from it. But don’t forget, classes only exist to create an object. To be able to create an object from an abstract class, we must create a new class that inherits from the superclass (base/parent).You can do this using the ```extends``` keyword.
+ 
+Here is a Code snippet to illustrate the ```extends``` keyword:
+ 
+```java
+ 
+public class MainDog extends Dog {
+ 
+ //abstract method body definition
+ 
+ public void run(){
```

Comment:
I would recommend adding the \@Override annotation on the run method

Reply:
Done